File: app/src/main/java/danb/speedrunbrowser/GameSubscribeDialog.java
package danb.speedrunbrowser;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.res.ColorStateList;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import java.util.List;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.NonNull;
import danb.speedrunbrowser.api.objects.Category;
import danb.speedrunbrowser.api.objects.Game;
import danb.speedrunbrowser.api.objects.Level;
import danb.speedrunbrowser.api.objects.Variable;
import danb.speedrunbrowser.utils.AppDatabase;

public class GameSubscribeDialog extends AlertDialog implements CompoundButton.OnCheckedChangeListener, View.OnClickListener {

    private Game mGame;
    private GameDetailFragment.GameSubscription mSubscribed;

    private Button mSelectAllButton;
    private Button mDeselectAllButton;
    private LinearLayout mCategoryCheckboxDisplay;
    private Button mCancelButton;
    private Button mApplyButton;

    public GameSubscribeDialog(Context ctx, @NonNull GameDetailFragment.GameSubscription sub) {
        super(ctx, AlertDialog.THEME_DEVICE_DEFAULT_DARK);

        mGame = sub.getGame();
        mSubscribed = sub;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        View rootView = ((LayoutInflater)getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(R.layout.dialog_subscribe_game, null);

        mSelectAllButton = rootView.findViewById(R.id.btnSelectAll);
        mDeselectAllButton = rootView.findViewById(R.id.btnSelectNone);
        mCategoryCheckboxDisplay = rootView.findViewById(R.id.layoutListCategory);
        mCancelButton = rootView.findViewById(R.id.btnCancel);
        mApplyButton = rootView.findViewById(R.id.btnSave);

        mSelectAllButton.setOnClickListener(this);
        mDeselectAllButton.setOnClickListener(this);
        mCancelButton.setOnClickListener(this);
        mApplyButton.setOnClickListener(this);

        setContentView(rootView);

        setViewData();
    }

    private void setViewData() {
        mCategoryCheckboxDisplay.removeAllViews();

        for(Category category : mGame.getCategories()) {
            if(category.getType().equals("per-level")) {
                for(Level level : mGame.getLevels()) {
                    String id = category.getId() + "_" + level.getId();
                    CheckBox cb = makeStyledCheckbox();
                    cb.setText(getContext().getResources().getString(R.string.render_level_category, category.getName(), level.getName()));
                    cb.setChecked(mSubscribed.contains(id));
                    cb.setTag(id);

                    mCategoryCheckboxDisplay.addView(cb);
                }
            }
            else {
                CheckBox cb = makeStyledCheckbox();
                cb.setText(category.getName());
                cb.setChecked(mSubscribed.contains(category.getId()));
                cb.setTag(category.getId());

                mCategoryCheckboxDisplay.addView(cb);
            }
        }
    }

    private CheckBox makeStyledCheckbox() {
        CheckBox checkbox = new CheckBox(getContext());
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            checkbox.setButtonTintList(getContext().getResources().getColorStateList(R.color.all_white));
        }

        checkbox.setTextColor(getContext().getResources().getColor(android.R.color.white));
        checkbox.setOnCheckedChangeListener(this);

        return checkbox;
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        String id = ((String)buttonView.getTag());

        if(isChecked)
            mSubscribed.add(id);
        else
            mSubscribed.remove(id);
    }

    @Override
    public void onClick(View v) {

        if(v == mSelectAllButton) {
            for(Category category : mGame.getCategories()) {
                if(category.getType().equals("per-level")) {
                    for(Level level : mGame.getLevels()) {
                        mSubscribed.add(category.getId() + "_" + level.getId());
                    }
                }
                else {
                    mSubscribed.add(category.getId());
                }
            }

            setViewData();
        }
        else if(v == mDeselectAllButton) {
            mSubscribed.clear();

            setViewData();
        }
        else if(v == mApplyButton) {
            dismiss();
        }
        else if(v == mCancelButton) {
            cancel();
        }
    }

    public GameDetailFragment.GameSubscription getSubscriptions() {
        return mSubscribed;
    }
}

File: app/src/main/java/danb/speedrunbrowser/GameSubscribeDialog.kt
@file:Suppress("DEPRECATION")

package danb.speedrunbrowser

import android.app.AlertDialog
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.LinearLayout
import androidx.core.content.ContextCompat

import danb.speedrunbrowser.api.objects.Game

class GameSubscribeDialog(ctx: Context, val subscriptions: GameDetailFragment.GameSubscription) : AlertDialog(ctx, if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) android.R.style.Theme_DeviceDefault_Dialog_Alert
    else THEME_DEVICE_DEFAULT_DARK), CompoundButton.OnCheckedChangeListener, View.OnClickListener {

    private val mGame: Game? = subscriptions.game

    private var mSelectAllButton: Button? = null
    private var mDeselectAllButton: Button? = null
    private var mCategoryCheckboxDisplay: LinearLayout? = null
    private var mCancelButton: Button? = null
    private var mApplyButton: Button? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val rootView = (context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater).inflate(R.layout.dialog_subscribe_game, null)

        mSelectAllButton = rootView.findViewById(R.id.btnSelectAll)
        mDeselectAllButton = rootView.findViewById(R.id.btnSelectNone)
        mCategoryCheckboxDisplay = rootView.findViewById(R.id.layoutListCategory)
        mCancelButton = rootView.findViewById(R.id.btnCancel)
        mApplyButton = rootView.findViewById(R.id.btnSave)

        mSelectAllButton!!.setOnClickListener(this)
        mDeselectAllButton!!.setOnClickListener(this)
        mCancelButton!!.setOnClickListener(this)
        mApplyButton!!.setOnClickListener(this)

        setContentView(rootView)

        setViewData()
    }

    private fun setViewData() {
        mCategoryCheckboxDisplay!!.removeAllViews()

        for ((id1, name, _, type) in mGame!!.categories!!) {
            if (type == "per-level") {
                for ((id2, name1) in mGame.levels!!) {
                    val id = id1 + "_" + id2
                    val cb = makeStyledCheckbox()
                    cb.text = context.resources.getString(R.string.render_level_category, name, name1)
                    cb.tag = id
                    cb.isChecked = subscriptions.contains(id)

                    mCategoryCheckboxDisplay!!.addView(cb)
                }
            } else {
                val cb = makeStyledCheckbox()
                cb.text = name
                cb.isChecked = subscriptions.contains(id1)
                cb.tag = id1

                mCategoryCheckboxDisplay!!.addView(cb)
            }
        }
    }

    private fun makeStyledCheckbox(): CheckBox {
        val checkbox = CheckBox(context)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            checkbox.buttonTintList = ContextCompat.getColorStateList(context, R.color.all_white)
        }

        checkbox.setTextColor(ContextCompat.getColor(context, android.R.color.white))
        checkbox.setOnCheckedChangeListener(this)

        return checkbox
    }

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        val id = buttonView.tag as String? ?: return

        if (isChecked)
            subscriptions.add(id)
        else
            subscriptions.remove(id)
    }

    override fun onClick(v: View) {

        if (v === mSelectAllButton) {
            for ((id, _, _, type) in mGame!!.categories!!) {
                if (type == "per-level") {
                    for ((id1) in mGame.levels!!) {
                        subscriptions.add(id + "_" + id1)
                    }
                } else {
                    subscriptions.add(id)
                }
            }

            setViewData()
        } else if (v === mDeselectAllButton) {
            subscriptions.clear()

            setViewData()
        } else if (v === mApplyButton) {
            dismiss()
        } else if (v === mCancelButton) {
            cancel()
        }
    }
}
